solutions/task_4.py

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python3
+"""
+TASK: Implement classes "Temperature", "Person" using Python descriptor protocols to execute unittests without errors.
+"""
+import re
+import unittest
+from datetime import datetime
+
+
+class Temperature(object):
+
+    def __init__(self, value=0):
+        self.fahrenheit = value
+
+    def get_celsius(self):
+        return (self.fahrenheit - 32)/1.8
+
+    def set_celsius(self, value):
+        self.fahrenheit = value*1.8 + 32
+
+    celsius = property(get_celsius, set_celsius)
+
+
+class Descriptor(object):
+
+    def __init__(self, attr_name, attr_type):
+        self.attr_name = attr_name
+        self.attr_type = attr_type
+
+    def __get__(self, instance, owner):
+        return instance.__dict__[self.attr_name]
+
+    def __set__(self, instance, attr_value):
+        if self.attr_type != type(attr_value):
+            raise TypeError
+        if self.attr_name == 'phone':
+            pattern = r"[0-9]{3} [0-9]{2} [0-9]{7}$"
+            if re.match(pattern, attr_value):
+                instance.__dict__.__setitem__(self.attr_name, "+{} ({}) {}-{}-{}".format(attr_value[:3],
+                                                                                         attr_value[4:6],
+                                                                                         attr_value[7:10],
+                                                                                         attr_value[10:12],
+                                                                                         attr_value[12:]))
+            else:
+                raise ValueError
+        else:
+            instance.__dict__.__setitem__(self.attr_name, attr_value)
```

Comment:
```instance.__dict__.__setitem__(self.attr_name, attr_value)```

I showed this code to these people:

![](https://memepedia.ru/wp-content/uploads/2017/06/%D1%82%D0%B8%D0%BC-%D1%80%D0%BE%D1%82.jpg)

Reply:
![](https://cs7.pikabu.ru/post_img/2014/06/15/6/1402820341_2026770426.jpg)

Reply:
![](http://301-1.ru/gen-mems/img_mems/849193e9e3e15a3409fffaa23b26c811.jpg)

Reply:
Seems that you've attached wrong pictures. They were definitely happy to face such a good example of coding) Sure they will use my experience in their work)

Reply:
Sure, they are crying due to happiness.